modules/imgproc/src/drawing_text.cpp

```diff
@@ -0,0 +1,1541 @@
+    hb_script_t script;
+    hb_direction_t dir;
+};
+
+struct FontGlyph
+{
+    FontGlyph() { ttface = 0; index = -1; scale = 1.f; }
+    FontGlyph(font_t* ttface_, int glyph_index_, const hb_glyph_position_t& pos_, float scale_)
+    {
+        ttface = ttface_;
+        index = glyph_index_;
+        pos = pos_;
+        scale = scale_;
+    }
+    font_t* ttface;
+    int index;
+    float scale;
+    hb_glyph_position_t pos;
+};
+
+
+class FontRenderEngine
+{
+public:
+    enum { MAX_CACHED_GLYPH_SIZE = 128, MAX_CACHE_SIZE=2048 };
+    FontRenderEngine()
+    {
+        hb_buf = hb_buffer_create();
+        hb_buffer_guess_segment_properties(hb_buf);
+        hb_uni_funcs = hb_unicode_funcs_get_default();
+        max_cache_size = (size_t)MAX_CACHE_SIZE;
+        glyph_buf = 0;
+        glyph_bufsz = 0;
+    }
+
+    ~FontRenderEngine()
+    {
+        hb_buffer_destroy(hb_buf);
+        for(int i = 0; i < BUILTIN_FONTS_NUM; i++)
+            builtin_ffaces[i] = FontFace();
+        if (glyph_buf)
+            free(glyph_buf);
+        glyph_buf = 0;
+        glyph_bufsz = 0;
+    }
+
+    void addToCache(const GlyphCacheKey& key, const GlyphCacheVal& val)
+    {
+        if(glyph_cache.size() == max_cache_size)
+        {
+            GlyphCacheEntryIt last = all_cached.end();
+            --last;
+            glyph_cache.erase(last->first);
+            all_cached.pop_back();
+        }
+
+        all_cached.push_front(std::make_pair(key, val));
+        GlyphCacheEntryIt first = all_cached.begin();
+        glyph_cache.insert(std::make_pair(first->first, first));
+    }
+
+    GlyphCacheVal* findCachedGlyph(const GlyphCacheKey& key)
+    {
+        auto it = glyph_cache.find(key);
+        if(it == glyph_cache.end())
+            return 0;
+        all_cached.splice(all_cached.begin(), all_cached, it->second);
+        return &it->second->second;
+    }
+
+    FontFace& getStdFontFace(int i)
+    {
+        CV_Assert(i >= 0 && i < BUILTIN_FONTS_NUM);
+        builtin_ffaces[i]->setStd(builtinFontData[i]);
+        return builtin_ffaces[i];
+    }
+
+    Point putText_( Mat& img, Size imgsize, const String& str_, Point org,
+                    const uchar* color, FontFace& fontface, double size,
+                    double weight, int flags, Range wrapRange, bool render,
+                    Rect* bbox );
+
+protected:
+    FontFace builtin_ffaces[BUILTIN_FONTS_NUM];
+    bool builtin_ffaces_initialized;
+
+    hb_unicode_funcs_t* hb_uni_funcs;
+
+    // LRU cache of glyphs
+    GlyphCacheVal new_cached;
+    std::list<GlyphCacheEntry> all_cached;
+    std::unordered_map<GlyphCacheKey, GlyphCacheEntryIt, GlyphCacheHash> glyph_cache;
+    size_t max_cache_size;
+
+    hb_buffer_t* hb_buf;
+    std::vector<unsigned> u32buf;
+    std::vector<TextSegment> segments;
+    std::vector<FontGlyph> glyphs;
+    std::vector<uchar> pixbuf;
+    uchar* glyph_buf;
+    int glyph_bufsz;
+};
+
+thread_local FontRenderEngine fontRenderEngine;
```

Comment:
Why is thread local?

Reply:
The ultimate goal (not achieved yet, because it requires further refactoring of stb_truetype) is to eliminate all memory allocations in putText() and also avoid necessity to put any locks when accessing fonts, including built-in fonts etc. Freetype (that may be eventually added as an option) is not thread-safe, stb_truetype is not thread-safe, Harfbuzz (to be added in some of the subsequent pullrequests) is not thread-safe. We embrace them, along with the glyph cache, various buffers etc. into a thread-local variable to make everything thread-safe.